### PR TITLE
Support for naming scrutinee in dependent pattern matching

### PR DESCRIPTION
**Pull request by aseemr**
### Dependent pattern matching in F*

PR #2256 added support for dependent pattern matching with `returns` annotations on `match`. E.g., from `ulib/Steel.ST.EphemeralHashtbl.fst`:

```
let vopt = Seq.index s (U32.v idx) in
match vopt returns STGhostT _ _ _ (fun _ -> tperm a (Map.upd m i c) (Map.remove borrows i)) with ...
```

However, it comes with a restriction that the scrutinee expression has to be a name in the presence of a `returns` annotation. As a result, the snippet above has to let-bind the scrutinee expression.

### This PR

This PR adds support for naming the scrutinee using `as` keyword similar to Coq. E.g. the code above can now be written as:

```
match Seq.index s (U32.v idx) as y returns STGhostT _ _ _ (fun _ -> tperm a (Map.upd m i c) (Map.remove borrows i)) with ...
```

where `y` may appear free in the type/computation type after `returns`.

As an optimization, note that the computation type after `returns` in the first snippet does not contain `vopt`, so in this case, we can even write:

```
match Seq.index s (U32.v idx) returns STGhostT _ _ _ (fun _ -> tperm a (Map.upd m i c) (Map.remove borrows i)) with ...
```

thereby eliminating the unnecessary naming of the scrutinee.

#### AST, syntax, desugaring, resugaring, reflection

In the parser AST, a returns annotation may now optionally have a `ident`. Whereas in the core syntax, the returns annotation is of the form `(b, asc)`, where `b` is a binder and `asc` is an `Syntax.ascription` node.

Parser AST returns annotation `(x, t)` desugars to `(x_b, asc)`, where `x_b` is a binder for `x` and `t` desugars to `asc`, may be containing the bound variable `x_b`.

On the other hand, parser AST returns annotation `t` desugars to `(uu___ret_, asc)`, i.e. we introduce a new binder. One subtlety here is that, it may be the case that source program does not name the scrutinee, but scrutinee is already a name and occurs free in the returns annotation, e.g.:

```
match x returns Tot (n:int{n > x}) with ...
```
In this case, when we introduce a new binder `uu___ret_`, we also substitute `x` in `t` with `uu___ret_`.

That way, core typechecker uniformly works with an explicit binder for the `returns` annotation.

When resugaring, we check that if the binder name for the returns annotation is the internal name `uu___ret_`, we do not print it, and appropriately substitute the scrutinee back if necessary. See https://github.com/FStarLang/FStar/blob/aseem_dependent_pattern_matching2/tests/prettyprinting/Match.Returns.fst and https://github.com/FStarLang/FStar/blob/aseem_dependent_pattern_matching2/tests/prettyprinting/Match.Returns.fst.expected for pretty-printing testcases.

Reflection AST, embeddings, and unembeddings are also fixed accordingly.

#### Typechecker

To typecheck

```
match e as x returns C with
| P_i -> e_i
```

We first typecheck `G |- e : C_e`. Then we typecheck `G, x : C_e.t |- C : _`, i.e. extend the environment with `x:t` where `t` is the type of `e` and typecheck `C`. Then we ascribe each branch as `e_i <: C[P_i/x]`, and typecheck the ascribed branch. The final type of the `match` result is computed as `C[e/x]` (as usual we later bind it with `C_e`).

### CODE REVIEW

File: src/typechecker/FStar.TypeChecker.TcTerm.fs [hunk 1]
```diff
@@ -1205,91 +1212,121 @@ and tc_match (env : Env.env) (top : term) : term * lcomp * guard_t =
...
-
-        let cases = List.map
-          (fun (f, eff_label, cflags, c) ->
-             (f, eff_label, cflags, (fun b -> TcComm.set_result_typ_lc (c b) t))) cases in
+      match ret_opt with
+      | Some (b, (Inr c, _)) ->  //a return annotation, with computation type
 
-        TcUtil.bind_cases env t cases guard_x, g, erasable
+        //c has b free, so substitute it with the scrutinee
+        let c = SS.subst_comp [NT (b.binder_bv, e1)] c in
```
Comment (aseemr) [inline src/typechecker/FStar.TypeChecker.TcTerm.fs:1268]: This is where we compute the type of the `match` result by substituting binder with `e1`.

File: src/typechecker/FStar.TypeChecker.TcTerm.fs [hunk 2]
```diff
@@ -1205,91 +1212,121 @@ and tc_match (env : Env.env) (top : term) : term * lcomp * guard_t =
...
-           | _ ->
-             raise_error (Errors.Fatal_UnexpectedTerm,
-               "The scrutinee must be a variable when a return annotation is supplied with a match") e1.pos) in
-      guard_x,
-      eqns |> List.map (tc_eqn guard_x env_branches ret_opt) in
+        let b, asc =
+          let bs, asc = SS.open_ascription [b] asc in
+          let b = List.hd bs in
+          //we set the sort of the binder to be the type of e1
+          {b with binder_bv={b.binder_bv with sort=c1.res_typ}}, asc in
```
Comment (aseemr) [inline src/typechecker/FStar.TypeChecker.TcTerm.fs:1235]: Adding the returns binder with sort as the type of `e1`, to typecheck the ascription.

File: src/typechecker/FStar.TypeChecker.TcTerm.fs [hunk 3]
```diff
@@ -1302,9 +1339,18 @@ and tc_match (env : Env.env) (top : term) : term * lcomp * guard_t =
...
+      //repack the returns ascription
+      //we make the binder sort as tun,
+      //  since we always use the type of the scrutinee
+      let ret_opt =
+        match ret_opt with
+        | None -> None
+        | Some (b, asc) ->
+          let asc = SS.close_ascription [b] asc in
+          let b = List.hd (SS.close_binders [b]) in
+          let b = {b with binder_bv={b.binder_bv with sort=tun}} in
```
Comment (aseemr) [inline src/typechecker/FStar.TypeChecker.TcTerm.fs:1351]: While packing the returns annotation back, set the sort to `_`. So we don't have to normalize it, instantiate it, compute its fvs etc. etc.

File: src/typechecker/FStar.TypeChecker.TcTerm.fs [hunk 4]
```diff
@@ -2948,27 +2998,29 @@ and tc_eqn scrutinee env ret_opt branch
...
   (* 3. Check the branch *)
   let branch_exp, c, g_branch =
     let branch_exp =  //ascribe with the return annotation, if it exists
       match ret_opt with
       | None -> branch_exp
-      | Some asc ->
+      | Some (b, asc) ->
         asc
-        |> SS.subst_ascription [NT (scrutinee, norm_pat_exp)]
+        |> SS.subst_ascription [NT (b.binder_bv, norm_pat_exp)]
```
Comment (aseemr) [inline src/typechecker/FStar.TypeChecker.TcTerm.fs:3017]: Ascribe the branches before checking them.

File: src/parser/FStar.Parser.Lexhelp.fs [hunk 5]
```diff
@@ -173,6 +173,7 @@ let keywords =
     ALWAYS, "let"        ,LET(false);
     ALWAYS, "logic"      ,LOGIC;
     ALWAYS, "match"      ,MATCH;
+    ALWAYS, "as"         ,AS;
```
Comment (nikswamy) [inline src/parser/FStar.Parser.Lexhelp.fs:176]: super tiny nit, sorry. But this list of keywords was originally meant to be in alphabetical order. I see there are some deviations from that already, but may be worth restoring at some point.

File: src/parser/FStar.Parser.ToDocument.fs [hunk 6]
```diff
@@ -1308,10 +1312,14 @@ and p_noSeqTerm' ps pb e = match e.tm with
...
-       | Some ret ->
-        group (surround 2 1 (str "match")
-                            ((p_noSeqTermAndComment false false e) ^/+^ (str "returns" ^/+^ p_tmIff ret))
-                            (str "with")))
+       | Some (as_opt, ret) ->
+         group (surround 2 1 (str "match")
+                             ((p_noSeqTermAndComment false false e) ^/+^
+                              (match as_opt with
+                               | None -> empty
+                               | Some as_ident -> str " as" ^/+^ (p_ident as_ident)) ^/+^
```
Comment (nikswamy) [inline src/parser/FStar.Parser.ToDocument.fs:1320]: do we need the preceding " " in ` as`?

File: src/parser/FStar.Parser.ToDocument.fs [hunk 7]
```diff
@@ -1292,10 +1292,14 @@ and p_noSeqTerm' ps pb e = match e.tm with
...
                (str "then" ^/+^ e2_doc) ^/^
                (str "else" ^/+^ p_noSeqTermAndComment ps pb e3))
-           | Some ret ->
+           | Some (as_opt, ret) ->
               group (
                 (str "if" ^/+^ p_noSeqTermAndComment false false e1) ^/^
-                (str "ret" ^/+^ p_tmIff ret) ^/^
+                ((match as_opt with
+                  | None -> empty
+                  | Some as_ident -> str " as " ^/^ p_ident as_ident ^/^ str " ")
```
Comment (nikswamy) [inline src/parser/FStar.Parser.ToDocument.fs:1300]: same as below ... do we need the spaces around it?

File: tests/prettyprinting/Match.Returns.fst.expected [hunk 8]
```diff
@@ -1,22 +1,32 @@
 module Match.Returns
 
 let test (n: nat) =
-  match n returns nat with
+  match n  returns nat with
```
Comment (nikswamy) [inline tests/prettyprinting/Match.Returns.fst.expected:4]: I wonder if this extra space will disappear with some tuning in ToDocument

File: ulib/LowStar.BufferView.Down.fst [hunk 9]
```diff
@@ -78,12 +78,12 @@ let indexing' (#a #b: _) (v:view a b) (len_as:nat) (i:nat{i < len_as * View?.n v
 let indexing #b vb i = indexing' (get_view vb) (B.length (as_buffer vb)) i
 
 let sel' (#a #b: _) (v:view a b)
-         (as:Seq.seq a)
-         (i:nat{i / View?.n v < Seq.length as})
+         (es:Seq.seq a)
```
Comment (nikswamy) [inline ulib/LowStar.BufferView.Down.fst:81]: thanks for all the fixes in this file!

File: ulib/experimental/Steel.ST.EphemeralHashtbl.fst [hunk 10]
```diff
@@ -852,11 +852,11 @@ let ghost_put #_ #k #v #contents #vp #h #m #borrows a i x c =
 
   //reestablish the invariant
 
-  let vopt = Seq.index s (U32.v idx) in
-  match vopt returns STGhostT _
-                              _
-                              _
-                              (fun _ -> tperm a (Map.upd m i c) (Map.remove borrows i)) with
+  match Seq.index s (U32.v idx)
```
Comment (nikswamy) [inline ulib/experimental/Steel.ST.EphemeralHashtbl.fst:855]: nice

File: src/typechecker/FStar.TypeChecker.TcTerm.fs (hunk 3 above)
Comment (nikswamy) [reply src/typechecker/FStar.TypeChecker.TcTerm.fs:1367]: interesting ... nice trick

**Comment by nikswamy** [comment]
Thanks Aseem.. This looks great! Thanks for all the explanations and code comments in TcTerm too. 

I left a few minor nit-picky comments.

File: src/parser/FStar.Parser.ToDocument.fs (hunk 6 above)
Comment (aseemr) [reply src/parser/FStar.Parser.ToDocument.fs:1320]: Thanks for noticing! I removed these extra spaces and updated the pretty printing tests.

File: ulib/FStar.Tactics.Print.fst [hunk 11]
```diff
@@ -37,24 +37,31 @@ let rec term_to_ast_string (t:term) : Tac string =
...
 
+and match_returns_to_string (ret_opt:option match_returns_ascription) : Tac string =
+  let tacopt_to_string tacopt : Tac string =
+    match tacopt with
+    | None -> ""
+    | Some tac -> " by " ^ (term_to_ast_string tac) in
+  match ret_opt with
+  | None -> ""
+  | Some (b, asc) ->
+    (binder_to_string b ^ " ")
```
[the remark this answers is not in the document]
Comment (aseemr) [reply ulib/FStar.Tactics.Print.fst:59]: This is simply printing the AST, e.g. Tm_match is already printed as (scrutinee, branches), so I followed the same style (e.g. no `returns` as well).

**Comment by aseemr** [comment]
Thanks for the review Nik! I incorporated the comments, except the alphabetical sorting of keywords in the lexer(s). I will raise an enhancement issue for that.

**Pull request merged**